### DISCUSSION TITLE
Possible Bug? Configuring Base Directory via RENOVATE_BASE_DIR Overridden

### DISCUSSION THREAD

**Discussion by danhlee329** [renovatebot/renovate#27628] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.219.5

### If you're self-hosting Renovate, select which platform you are using.

github.com

### What is your question?

I'm testing Renovate locally via Docker and am finding a possible bug around the `base_dir` being overridden after init.

* I'm running the command below with the config set:
<details><summary>Docker Command and Renovate Config</summary>

```bash
BASE_DIR="/tmp/renovate/temp"
LOCAL_DIR="/local/dir"
docker run \
    --rm \
    -e LOG_LEVEL="debug" \
    -e RENOVATE_PLATFORM="github" \
    -e RENOVATE_TOKEN="$GH_TOKEN" \
    -e RENOVATE_BASE_DIR="$BASE_DIR" \
    -e RENOVATE_CUSTOM_ENV_VARIABLES="{\"TRACKER_FILE_PATH\": \"$BASE_DIR/module-tracker.json\"}" \
    -e RENOVATE_CONFIG_FILE="$BASE_DIR/renovate.json" \
    -v "$LOCAL_DIR:$BASE_DIR" \
    renovate/renovate:"37.219.5" \
        --dry-run="true" | tee -a renovate-$(date +"%Y%m%d%H%M").log
```

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  ...
  "customDatasources": {
    "local_module_tracker": {
      "defaultRegistryUrlTemplate": "file://{{ env.TRACKER_FILE_PATH }}",
      "format": "json",
      "transformTemplates": [
        "{ $map($, function($v) { { $v.depName: { \"lastModified\": $v.version } } }) }"
      ]
    }
  },
  "customManagers": [
    {
      "customType": "regex",
      "fileMatch": [".*.$EXT"],
      "datasourceTemplate": "custom.local_module_tracker",
      "matchStrings": [...]
    }
  ],
  "repositories": ["$OWNER/$REPO"]
}
```

</details>

* File-matching for the manager works, but can't seem to get past this error (for all matches), which makes sense when looking at [this function](https://github.com/renovatebot/renovate/blob/1efa2f21da4f4b2c0c984726c1a35038d76732db/lib/util/fs/util.ts#L6-L14) (in `Log 1` below)
<details><summary>Logs - 1</summary>

```
DEBUG: Preventing access to file outside the base directory (repository=$OWNER/$REPO)
       "path": "$BASE_DIR/module-tracker.json",
       "baseDir": "$BASE_DIR/repos/github/$OWNER/$REPO"
```

</details>

* When setting env var `BASE_DIR` to `/tmp/renovate/temp/repos/github/$OWNER/$REPO`, getting error below (in `Log 2` below)
<details><summary>Logs - 2</summary>

```
DEBUG: Git function thrown (repository=$OWNER/$REPO)
       "err": {
         "errno": -16,
         "code": "EBUSY",
         "syscall": "unlink",
         "path": "$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json",
         "message": "EBUSY: resource busy or locked, unlink '$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json'",
         "stack": "Error: EBUSY: resource busy or locked, unlink '$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json'"
       }
DEBUG: git clone error (repository=$OWNER/$REPO)
       "err": {
         "errno": -16,
         "code": "EBUSY",
         "syscall": "unlink",
         "path": "$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json",
         "message": "EBUSY: resource busy or locked, unlink '$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json'",
         "stack": "Error: EBUSY: resource busy or locked, unlink '$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json'"
       }
 WARN: Host error (repository=$OWNER/$REPO)
       "hostType": "git",
       "packageName": undefined,
       "err": {
         "errno": -16,
         "code": "EBUSY",
         "syscall": "unlink",
         "path": "$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json",
         "message": "EBUSY: resource busy or locked, unlink '$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json'",
         "stack": "Error: EBUSY: resource busy or locked, unlink '$BASE_DIR/repos/github/$OWNER/$REPO/module-tracker.json'"
       }
 INFO: External host error causing abort - skipping (repository=$OWNER/$REPO)
```

</details>

The `module-tracker-json` file is something that requires us to download this from a different source and have it available when we run `renovate` thus requiring us to use the `file://` prefix per [Custom offline dependencies](https://docs.renovatebot.com/modules/datasource/custom/#custom-offline-dependencies).  The question is whether the `baseDir` should be overridden if we provide one during init.  Because we're providing a `RENOVATE_BASE_DIR`, I wouldn't expect the `baseDir` to change.

**Comment by rarkins** [comment]
Base dir needs to be a directory where Renovate is free to create its own directory structure within. You've not given enough logs to be sure, but it seems to be behaving as designed.

Why are you changing the base dir in the first place - what's your requirement?

**Reply by danhlee329** [reply]
**Here are some more logs that has more context around it behaving 'as designed', but misleading with the log under `Logs - 1` above:**

* For context, `RENOVATE_BASE_DIR` is set to `/tmp/renovate/temp` in the original call:

<details>
<summary>Relevant Debug Logs</summary>

```
...
DEBUG: Using configured baseDir: /tmp/renovate/temp
DEBUG: Using cacheDir: /tmp/renovate/temp/cache
DEBUG: Using containerbaseDir: /tmp/renovate/temp/cache/containerbase
DEBUG: Initializing Renovate internal cache into /tmp/renovate/temp/cache/renovate/renovate-cache-v1
DEBUG: Commits limit = null
DEBUG: Setting global hostRules
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
DEBUG: validatePresets()
DEBUG: Reinitializing hostRules for repo
DEBUG: Clearing hostRules
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
 INFO: Repository started (repository=$OWNER/$REPO)
       "renovateVersion": "37.219.5"
DEBUG: Using localDir: /tmp/renovate/temp/repos/github/$OWNER/$REPO (repository=$OWNER/$REPO)
DEBUG: PackageFiles.clear() - Package files deleted (repository=$OWNER/$REPO)
DEBUG: initRepo("$OWNER/$REPO") (repository=$OWNER/$REPO)
DEBUG: hostRules: authentication already set for api.github.com (repository=$OWNER/$REPO)
...
DEBUG: Initializing git repository into /tmp/renovate/temp/repos/github/$OWNER/$REPO (repository=$OWNER/$REPO)
...
DEBUG: Preventing access to file outside the base directory (repository=$OWNER/$REPO)
       "path": "/tmp/renovate/temp/module-tracker.json",
       "baseDir": "/tmp/renovate/temp/repos/github/$OWNER/$REPO"
...
```

</details>

**Here are some notes around our particular use-case pushing us to customize the `baseDir`:**
- <ins>We currently have a version-file that holds all-latest-updates for custom modules/packages within a separate Github repo (saved as an asset in a release).  In the logs/commands above, you'll see this file named as `module-tracker.json`</ins>
- <ins>Because we want to utilize `module-tracker.json` as our main-source-of-truth for current-versions for our modules/packages, [custom offline dependencies](https://docs.renovatebot.com/modules/datasource/custom/#custom-offline-dependencies) supports our use-case the best.  Because we need to download this file (as an asset) from a particular release, we'll want to utilize the `file://` prefix in the `customDatasources.$DS_NAME. defaultRegistryUrlTemplate` in the config once we download the asset.</ins>
  - When [getting a Github release asset](https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28), looks like there are a couple ways we can get asset info or the asset itself (tried both as the URL, but not working via Renovate):
    - Asset Info (got a `200` for this request, but it's json on the asset's metadata, not the asset itself): `https://api.github.com/repos/$OWNER/$REPO/releases/assets/$ASSET_ID`
    - Actual Asset (works via browser, not via Renovate, getting a `404`): `https://github.com/$OWNER/$REPO/releases/download/$TAG_ID/module-tracker.json`
- <ins>We're able to download the asset before we run self-hosted Renovate. We'll **should** be able to access the file locally by using the `file://` prefix and mount it to the running Renovate container instance (either locally or via Github Actions)</ins>

While the `baseDir` is the directory where Renovate creates it's own directory structure, it's confusing with the `Preventing access to file outside the base directory` error since the `baseDir` listed is the directory of the cloned Github repo (in the logs, looks like this is set as the `localDir`).  Perhaps this needs to be renamed to something like `relativeBaseDir` instead of `baseDir`.

If we use the `file://` prefix, do we assume that the file is already in the Github repo that we clone in the first place or is this a bug where we've assumed that we can access any file set within a `baseDir` and it's sub-directories?

**Reply by rarkins** [reply, marked as answer]
It's assumed that the file is already in the repo

**Reply by danhlee329** [reply]
@rarkins thanks for the clarification.  While this answers the question on hand, I'll plan to submit an enhancement request to see if we can modify docs and var-names in the code for better clarification.

**Comment by danhlee329** [comment]
Follow up discussion/enhancment-request here: https://github.com/renovatebot/renovate/discussions/27644
